pandas/core/indexes/interval.py

```diff
@@ -1094,28 +1130,78 @@ def equals(self, other):
     def overlaps(self, other):
         return self._data.overlaps(other)
 
-    def _setop(op_name, sort=None):
-        def func(self, other, sort=sort):
-            self._assert_can_do_setop(other)
-            other = ensure_index(other)
-            if not isinstance(other, IntervalIndex):
-                result = getattr(self.astype(object), op_name)(other)
-                if op_name in ('difference',):
-                    result = result.astype(self.dtype)
-                return result
-            elif self.closed != other.closed:
-                msg = ('can only do set operations between two IntervalIndex '
-                       'objects that are closed on the same side')
-                raise ValueError(msg)
+    @Appender(_index_shared_docs['intersection'])
+    @SetopCheck(op_name='intersection')
+    def intersection(self, other, sort=False):
+        if self.left.is_unique and self.right.is_unique:
+            taken = self._intersection_unique(other)
+        else:
+            # duplicates
+            taken = self._intersection_non_unique(other)
 
-            # GH 19016: ensure set op will not return a prohibited dtype
-            subtypes = [self.dtype.subtype, other.dtype.subtype]
-            common_subtype = find_common_type(subtypes)
-            if is_object_dtype(common_subtype):
-                msg = ('can only do {op} between two IntervalIndex '
-                       'objects that have compatible dtypes')
-                raise TypeError(msg.format(op=op_name))
+        if sort is None:
+            taken = taken.sort_values()
+
+        return taken
+
+    def _intersection_unique(self, other):
+        """
+        Used when the IntervalIndex does not have any common endpoint,
+        no mater left or right.
+        Return the intersection with another IntervalIndex.
+
+        Parameters
+        ----------
+        other : IntervalIndex
+
+        Returns
+        -------
+        taken : IntervalIndex
+        """
+        lindexer = self.left.get_indexer(other.left)
+        rindexer = self.right.get_indexer(other.right)
+
+        match = (lindexer == rindexer) & (lindexer != -1)
+        indexer = lindexer.take(match.nonzero()[0])
+
+        return self.take(indexer)
 
+    def _intersection_non_unique(self, other):
+        """
+        Used when the IntervalIndex does have some common endpoints,
+        on either sides.
+        Return the intersection with another IntervalIndex.
+
+        Parameters
+        ----------
+        other : IntervalIndex
+
+        Returns
+        -------
+        taken : IntervalIndex
+        """
+        mask = np.zeros(len(self), dtype=bool)
```

Comment:
There might be an issue with this approach when dupes are present in `self` and `other`.  For other index types, such a scenario can result in more dupes being present in the intersection than in `self`.  This behavior looks a bit buggy and inconsistent though, so I'm not sure if we actually want `IntervalIndex` to be consistent with it.

Some examples of the buggy and inconsistent behavior with `Index`:
```python
In [2]: idx2 = pd.Index(list('aa')) 
   ...: idx3 = pd.Index(list('aaa')) 
   ...: idx3b = pd.Index(list('baaa'))

In [3]: idx2.intersection(idx3)
Out[3]: Index(['a', 'a', 'a', 'a'], dtype='object')

In [4]: idx3.intersection(idx3)
Out[4]: Index(['a', 'a', 'a'], dtype='object')

In [5]: idx2.intersection(idx3)
Out[5]: Index(['a', 'a', 'a', 'a'], dtype='object')

In [6]: idx2.intersection(idx3b)
Out[6]: Index(['a', 'a', 'a'], dtype='object')
```
It seems strange that `[3]` has more dupes present than in either original index but `[4]` does not.  Similarly, it seems like `[5]` and `[6]` should be identical, as the presence of a non-intersecting element shouldn't impact the number of dupes returned.

@jreback : Do you know what the expected behavior for `intersection` with dupes should be?  Or if there are any dependencies on the behavior of `intersection` that would dictate this?

If we treat indexes like multisets, then the intersection should contain the minimum multiplicity of dupes, e.g. `idx2.intersection(idx3)` and `idx3.intersection(idx2)` should both have length 2, so you maintain the property of the intersection being a subset of the original indexes.

Reply:
yeah this is weird as these are set ops

what happens (meaning how much breakage) if
- raise if left and right are not unique
- uniquify left and right

prob need to do this for all set ops

Reply:
Haven't had time to extensively test this out but I made the two changes you suggested in `indexes/base.py` for `intersection` and both resulted in some breakage.  Aside some from breakage in the index set ops tests, there was also some breakage in `tests/reshape/test_merge.py`.